### PR TITLE
add nansum

### PR DESCRIPTION
**Pull request by marigoold**
概述：实现了 oneflow.nansum 算子
实现：
- 和 reduce_sum 思路类似，但是增加了一个 binary_func: nansum
- backward 和 reduce_sum 相同，都是 ones_like
- 原来的 reduce 系列算子中，当某个需要 reduce 的 dim 上的 size 为 1 时，会自动进行 memcpy 操作。这里和 nansum 的结果冲突（因为要把 nan 转成 0），泽康实现了一个 primitive 的 unary func 用来处理 nan

注意事项：
- 暂未支持 FP16

文档：
image

### CODE REVIEW

File: oneflow/core/ndarray/xpu_ndarray_assign.h [hunk 1]
```diff
@@ -45,6 +47,17 @@ struct XpuNdarrayAssign<
     if (x.ptr() == y.ptr()) { return; }
     Memcpy<device_type>(stream, y.ptr(), x.ptr(), y.shape().ElemNum() * sizeof(T));
   }
+
+  static void AssignNanSum(ep::Stream* stream, const XpuVarNdarray<T>& y,
+                           const XpuVarNdarray<const T>& x) {
+    CHECK(y.shape() == x.shape());
```
Comment (BBuf) [inline oneflow/core/ndarray/xpu_ndarray_assign.h:53]: 这里可以给一个提示或者nolint

**Review by BBuf — APPROVED** [review]

File: oneflow/core/ndarray/xpu_ndarray_assign.h (hunk 1 above)
Comment (marigoold) [reply oneflow/core/ndarray/xpu_ndarray_assign.h]: 已改

File: oneflow/ir/include/OneFlow/OneFlowUserOps.td [hunk 2]
```diff
@@ -7068,6 +7068,23 @@ def OneFlow_ReduceSumOp : OneFlow_BaseOp<"reduce_sum", [NoSideEffect, DeclareOpI
   let has_data_type_infer_fn = 1;
 }
 
+def OneFlow_ReduceNanSumOp : OneFlow_BaseOp<"reduce_nansum", [NoSideEffect, DeclareOpInterfaceMethods<UserOpCompatibleInterface>]> {
```
[the remark this answers is not in the document]
Comment (marigoold) [reply oneflow/ir/include/OneFlow/OneFlowUserOps.td:7071]: 已改

File: oneflow/core/functional/impl/math_functor.cpp [hunk 3]
```diff
@@ -466,6 +460,64 @@ class ReduceSumFunctor {
...
+        one::OpBuilder("reduce_nansum").Input("input_tensor").Output("output_tensor").Build());
+  }
+  Maybe<Tensor> operator()(const std::shared_ptr<one::Tensor>& x,
+                           const Optional<Symbol<DType>>& dtype) const {
+    std::shared_ptr<one::Tensor> tensor = x;
+    if (dtype.has_value() && (dtype != x->dtype())) {
+      tensor = JUST(Cast(x, JUST(dtype), /*pin_memory=*/false));
+    }
+
+    const int32_t naxis = tensor->ndim();
```
[the remark this answers is not in the document]
Comment (marigoold) [reply oneflow/core/functional/impl/math_functor.cpp]: 已改

File: oneflow/core/ndarray/binary_func.h [hunk 4]
```diff
@@ -290,6 +300,14 @@ struct BinaryFuncAdd<half> final {
   static __device__ __forceinline__ half Invoke(const half x, const half y) { return __hadd(x, y); }
 };
 
+template<>
+struct BinaryFuncNanSum<half> final {
+  static __device__ __forceinline__ half Invoke(const half x, const half y) {
+    if (std::isnan(__half2float(x))) return std::isnan(__half2float(y)) ? __double2half(0.) : y;
```
Comment (MARD1NO) [inline oneflow/core/ndarray/binary_func.h:306]: 这里half类型就只会有cuda了，不用std库的isnan

File: oneflow/core/ndarray/binary_func.h [hunk 5]
```diff
@@ -66,6 +67,15 @@ struct BinaryFuncTrait final {
     }                                                                                         \
   }
 
+template<typename T>
+struct BinaryFuncNanSum final {
+  static OF_DEVICE_FUNC T Invoke(const T x, const T y) {
+    if (std::isnan(x)) return std::isnan(y) ? T{0} : y;
```
Comment (MARD1NO) [inline oneflow/core/ndarray/binary_func.h:73]: 这里要区分下，参考BinaryFuncFloorMod，用宏做条件编译

File: oneflow/core/ndarray/ndarray_reduce_impl.cu [hunk 6]
```diff
@@ -154,6 +164,7 @@ struct CubFunctor4BianryFunc;
     using type = cub::func_name;                                 \
   };
 OF_PP_FOR_EACH_ATOMIC(SPECIALIZE_CUB_FUNCTOR_4_BINARY_FUNC, REDUCE_BINARY_FUNC_NAME_SEQ);
+OF_PP_FOR_EACH_ATOMIC(SPECIALIZE_CUB_FUNCTOR_4_BINARY_FUNC, (NanSum));
```
Comment (MARD1NO) [inline oneflow/core/ndarray/ndarray_reduce_impl.cu:167]: 合并到166行吧

**Review by MARD1NO — CHANGES_REQUESTED** [review]

File: oneflow/core/ndarray/ndarray_reduce_impl.cu (hunk 6 above)
Comment (marigoold) [reply oneflow/core/ndarray/ndarray_reduce_impl.cu]: 已改

File: oneflow/core/ndarray/binary_func.h (hunk 5 above)
Comment (marigoold) [reply oneflow/core/ndarray/binary_func.h:77]: 已改

File: oneflow/core/ndarray/binary_func.h (hunk 4 above)
Comment (marigoold) [reply oneflow/core/ndarray/binary_func.h]: 已改

File: oneflow/core/ndarray/ndarray_reduce_impl.h [hunk 7]
```diff
@@ -16,6 +16,9 @@ limitations under the License.
 #ifndef ONEFLOW_CORE_NDARRAY_NDARRAY_REDUCE_IMPL_H_
 #define ONEFLOW_CORE_NDARRAY_NDARRAY_REDUCE_IMPL_H_
 
+#include "oneflow/core/common/data_type.h"
```
Comment (MARD1NO) [inline oneflow/core/ndarray/ndarray_reduce_impl.h:19]: 这三个头文件是不是没必要

File: python/oneflow/test/modules/test_global_nansum.py [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+@autotest(n=3, check_graph=False)
+def _test_global_nansum_with_0_size_tensor(test_case, placement, sbp):
+    x = random_tensor(4, 8, 16, 0, 24).to_global(placement, sbp)
+    mask = torch.ones_like(x).bool()
+    x = x.masked_fill(mask, float("nan"))
+    y = torch.nansum(x, dim=random(0, 3).to(int))
+    return y
+
+
+class TestGlobalSumModule(flow.unittest.TestCase):
```
Comment (MARD1NO) [inline python/oneflow/test/modules/test_global_nansum.py:46]: GlobalNanSum，下面的名字也要区别出来

**Review by MARD1NO — APPROVED** [review]
